pkg/distro/rhel/rhel_common/imagetype.go

```diff
@@ -0,0 +1,346 @@
+	// of the same name from the distro and arch
+	staticPackageSets := make(map[string]rpmmd.PackageSet)
+
+	for name, getter := range t.packageSets {
+		staticPackageSets[name] = getter(t)
+	}
+
+	// amend with repository information and collect payload repos
+	payloadRepos := make([]rpmmd.RepoConfig, 0)
+	for _, repo := range repos {
+		if len(repo.PackageSets) > 0 {
+			// only apply the repo to the listed package sets
+			for _, psName := range repo.PackageSets {
+				if slices.Contains(t.PayloadPackageSets(), psName) {
+					payloadRepos = append(payloadRepos, repo)
+				}
+				ps := staticPackageSets[psName]
+				ps.Repositories = append(ps.Repositories, repo)
+				staticPackageSets[psName] = ps
+			}
+		}
+	}
+
+	w := t.workload
+	if w == nil {
+		cw := &workload.Custom{
+			BaseWorkload: workload.BaseWorkload{
+				Repos: payloadRepos,
+			},
+			Packages: bp.GetPackagesEx(false),
+		}
+		if services := bp.Customizations.GetServices(); services != nil {
+			cw.Services = services.Enabled
+			cw.DisabledServices = services.Disabled
+		}
+		w = cw
+	}
+
+	containerSources := make([]container.SourceSpec, len(bp.Containers))
+	for idx, cont := range bp.Containers {
+		containerSources[idx] = container.SourceSpec{
+			Source:    cont.Source,
+			Name:      cont.Name,
+			TLSVerify: cont.TLSVerify,
+			Local:     cont.LocalStorage,
+		}
+	}
+
+	source := rand.NewSource(seed)
+	// math/rand is good enough in this case
+	/* #nosec G404 */
+	rng := rand.New(source)
+
+	img, err := t.image(w, t, bp.Customizations, options, staticPackageSets, containerSources, rng)
+	if err != nil {
+		return nil, nil, err
+	}
+	mf := manifest.New()
+
+	switch t.Arch().Distro().Releasever() {
+	case "9":
+		mf.Distro = manifest.DISTRO_EL9
+	case "10":
+		mf.Distro = manifest.DISTRO_EL10
+	default:
+		return nil, nil, fmt.Errorf("unsupported distro release version: %s", t.Arch().Distro().Releasever())
+	}
+
+	_, err = img.InstantiateManifest(&mf, repos, t.arch.distro.runner, rng)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &mf, warnings, err
+}
+
+// checkOptions checks the validity and compatibility of options and customizations for the image type.
+// Returns ([]string, error) where []string, if non-nil, will hold any generated warnings (e.g. deprecation notices).
+func (t *ImageType) checkOptions(bp *blueprint.Blueprint, options distro.ImageOptions) ([]string, error) {
+	if t.arch.distro.CheckOptions != nil {
+		return t.arch.distro.CheckOptions(t, bp, options)
+	}
+
+	return nil, nil
+}
+
+func NewImageType(
+	name, filename, mimeType string,
+	pkgSets map[string]PackageSetFunc,
+	imgFunc ImageFunc,
+	buildPipelines, payloadPipelines, exports []string,
+) *ImageType {
+	return &ImageType{
+		name:             name,
+		filename:         filename,
+		mimeType:         mimeType,
+		packageSets:      pkgSets,
+		image:            imgFunc,
+		buildPipelines:   buildPipelines,
+		payloadPipelines: payloadPipelines,
+		exports:          exports,
+	}
+}
```

Comment:
Is this function worth having?  I understand we might have it do more in the future, but right now, since the arguments map directly to the struct fields, it feels a bit redundant.  Even worse, it makes it easier to make mistakes with the position of args when calling the function (e.g. mixing up the last three args that are all `[]string`).

I suppose this way it makes it easier to control when adding new fields that are mandatory.  We could add a new argument and any caller that doesn't set it wouldn't compile.  Not sure what's better honestly.

This function is also consistent with our pipeline and stage generation code.  We add mandatory fields in the constructor and leave optional ones to be set by the caller.  So I guess it's better to keep and if we rethink that convention we can consider this one too.

Reply:
This is a very reasonable comment, and I was considering the same thing (having this function vs. having all struct fields exported and set directly).

You correctly deduced that one of the reasons for having this function is to make the mandatory fields required. IOW one must use the function to set them and must set all of them. There is exactly one image type which only calls this function and does not need to set any additional fields - `tar` image. Lastly, I also considered the consistency aspect with the pipeline and stage generation code. :slightly_smiling_face: 

But the main reason for this approach was the fact, that some struct fields can't be really exported with their current names due to the `distro.ImageType` interface methods names. This would mean that I would not name `name` as `Name`, because there is a `Name()` method in the interface. Unfortunately, the `buildPipelines`, `payloadPipelines` and `exports` have the same problem :disappointed:.  As a result, I would have to name some of the struct fields in a weird non-conflicting way, which didn't seem like the best idea.

Nevertheless, I'm open to suggestions.

Reply:
Ah yeah, I didn't consider the method names on the interface.  That's a good reason.  Let's leave it.  At least we have the consistency.